### PR TITLE
Fix ldexp backward zero gradient for negative integer exponents

### LINKED ISSUE #187460: torch.ldexp backward returns zero gradient w.r.t. mantissa for negative integer exponents
### 🐛 Describe the bug

`torch.ldexp(input, other)` computes `input * 2**other`, so the gradient w.r.t. `input` is `2**other`. When `other` is an integer tensor with a negative value, the backward pass returns **0** instead, while the forward pass is correct:

```python
import torch
m = torch.tensor([0.5], requires_grad=True)
e = torch.tensor([-1], dtype=torch.int32)
torch.ldexp(m, e).backward()
print(m.grad)   # tensor([0.])  — should be 0.5

# float exponent is correct:
m = torch.tensor([0.5], requires_grad=True)
torch.ldexp(m, torch.tensor([-1.0])).backward()
print(m.grad)   # tensor([0.5000])
```

The wrong gradient exactly equals `2**e` evaluated in **integer** arithmetic (floors to 0 for any negative `e`; correct for `e >= 0`), so the backward multiplier is being computed with an integer power instead of the real-valued `2.0**e` the forward uses:

| e (int32) | forward `ldexp(1, e)` | grad w.r.t. mantissa | correct `2**e` |
|---|---|---|---|
| -1 | 0.5 | **0.0** | 0.5 |
| -2 | 0.25 | **0.0** | 0.25 |
| -3 | 0.125 | **0.0** | 0.125 |
| 0 | 1.0 | 1.0 | 1.0 |
| 2 | 4.0 | 4.0 | 4.0 |

Finite differences confirm the true derivative: `(ldexp(0.5+h, -1) - ldexp(0.5-h, -1)) / 2h = 0.5`. Forward and backward therefore disagree.

**Impact:** silent — differentiating through `ldexp(x, e)` with a negative integer exponent yields a zero gradient, so the parameter simply stops training, with no error.

**Suggested fix:** cast `other` to the input's floating dtype before computing `2**other` in the backward formula (matching the forward).

### Versions

Reproduced on torch 2.12.0, CPU (arm64, macOS).

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

### PR DESCRIPTION
**Pull request by wasimat404**
Fixes #187460
torch.ldexp(input, other) computes input * 2**other, so the gradient w.r.t. input is 2**other. As @OE-GOD diagnosed in the issue, the backward formula in derivatives.yaml computed at::pow(2, other) with an integer-typed other, so 2**e was evaluated in integer arithmetic and floored to 0 for any negative e — silently zeroing the gradient while the forward pass stayed correct.
Fix: cast other to the result's floating dtype before the pow, matching the forward path:

`self: grad * at::pow(2, other.to(result.scalar_type())).conj()`

Test: added a regression test in test_autograd.py covering negative, zero, and positive integer exponents (-1, -2, 0, 2), asserting the mantissa gradient equals 2**e. The existing BinaryUfuncInfo gradcheck coverage never exercised a negative integer exponent with grad on the mantissa, which is why this went unnoticed.
I couldn't build locally (hardware constraints) so I'm relying on CI to verify. Happy to also add an OpInfo sample input for ldexp covering this case if reviewers prefer that over the standalone test.
cc @soulitzer @albanD

### CODE REVIEW

**Review by soulitzer — CHANGES_REQUESTED** [review]
This adds a unnecessary copy

**Comment by wasimat404** [comment]
> This adds a unnecessary copy

right. switched to a float base -- at::pow(2.0, other) , so it promotes to float without copying other. checked that pow(2.0, <int tensor>) gives the right values for negative exponents (-1 → 0.5 etc).
One thing I wanted to know: the jvp formula right below uses an int base too (at::pow(2, other_p)). could it have the same issue, or does the forward-mode path not hit that for integer exponents? I couldn't put together a forward-mode repro locally, so wasn't sure either way  happy to draw in a follow-up if it's worth it.

**Comment by soulitzer** [comment]
Thanks for the PR and the updates @wasimat404, but the fix is already in the process of landing now in https://github.com/pytorch/pytorch/pull/186566 actually. Going to close this one.
